Etcd Downgrade backend

YoyinZyc opened the pull request:
The original [downgrade PR](https://github.com/etcd-io/etcd/pull/11362) is too big to review. I will break it down into multiple small PRs.
This pr contains the server side of downgrade. The unit tests will be added in following PRs.

Based on the discussion [here](https://github.com/etcd-io/etcd/issues/11385), the cluster only allows member with downgrade target version to join in during downgrading. 
For more information about the design, please go to https://github.com/etcd-io/etcd/pull/11362
@gyuho @jingyih @wenjiaswe  @hexfusion PTAL Thanks:)

gyuho commented on etcdserver/api/membership/cluster.go:
```diff
@@ -778,6 +796,36 @@ func clusterVersionFromBackend(lg *zap.Logger, be backend.Backend) *semver.Versi
 
+func downgradeInfoFromBackend(lg *zap.Logger, be backend.Backend) *DowngradeInfo {
+	dkey := backendDowngradeKey()
+	if be != nil {
```
Can we panic on `be == nil`? So, we can reduce indentation?

YoyinZyc replied:
Move nil check outside `downgradeInfoFromBackend` to reduce indentation.

gyuho commented on etcdserver/api/membership/cluster.go:
```diff
@@ -778,6 +796,36 @@ func clusterVersionFromBackend(lg *zap.Logger, be backend.Backend) *semver.Versi
+		}
+
+		if len(keys) != 1 {
+			if lg != nil {
```
Also, do we really need `nil` checks on logger?

YoyinZyc replied:
I don't quite get it. Since we haven't deprecated plog and used zap log as default, lg is set to nil as default if flag `--logger=zap` is not enabled. I think we can remove all the nil check of log after we enable zap log as default?

gyuho replied:
I meant, if we set logger to https://godoc.org/go.uber.org/zap#NewNop once in the caller, all the downstream calls need not handle this `nil` checks?

gyuho commented on etcdserver/api/membership/cluster.go:
```diff
@@ -811,23 +859,70 @@ func ValidateClusterAndAssignIDs(lg *zap.Logger, local *RaftCluster, existing *R
+	if d != nil && d.Enabled && d.TargetVersion != nil {
+		if isValidDowngrade(d.TargetVersion, lv) {
+			if cv != nil {
+				if lg != nil {
```
Same. Can previous calls ensure logger is always non-nil?

gyuho commented on etcdserver/api/membership/cluster.go:
```diff
@@ -811,23 +859,70 @@ func ValidateClusterAndAssignIDs(lg *zap.Logger, local *RaftCluster, existing *R
 
+// isValidDowngrade checks whether joining local server is a valid downgrade when cluster enables downgrade.
+// the validation passes when local version is one minor version higher then target version
+func isValidDowngrade(tv *semver.Version, lv *semver.Version) bool {
```
this needs unit tests?

YoyinZyc replied:
Yes, it has unit test but not in this pr. I thought it will be a big pr if I include all the tests. Should I put them in or just add the unit tests?

gyuho commented on etcdserver/api/membership/store.go:
```diff
@@ -75,6 +76,20 @@ func mustSaveClusterVersionToBackend(be backend.Backend, ver *semver.Version) {
+	dkey := backendDowngradeKey()
+	dvalue, err := json.Marshal(downgrade)
+	if err != nil {
+		if lg != nil {
```
Also, can previous calls ensure logger is non-nil always?

gyuho commented on etcdserver/api/membership/store.go:
```diff
@@ -75,6 +76,20 @@ func mustSaveClusterVersionToBackend(be backend.Backend, ver *semver.Version) {
+	}
+	tx := be.BatchTx()
+	tx.Lock()
+	defer tx.Unlock()
```
unnecessary `defer`

YoyinZyc replied:
Done.

gyuho commented on etcdserver/cluster_util.go:
```diff
@@ -177,6 +177,21 @@ func getVersions(lg *zap.Logger, cl *membership.RaftCluster, local types.ID, rt
 	return vers
 }
 
+// allowedVersionRange decides the available version range of the cluster that local server can join in;
```
Can these be unit tested?

gyuho commented on etcdserver/server.go:
```diff
@@ -2617,6 +2678,39 @@ func (s *EtcdServer) updateClusterVersion(ver string) {
+
+		targetVersion := d.TargetVersion
+		if isDowngradeFinished(s.getLogger(), targetVersion, getVersions(s.getLogger(), s.cluster, s.id, s.peerRt)) {
+			if lg != nil {
```
Here as well, logger nil checking should be handled at the top level. Downstream packages should not worry about this checks.

jingyih commented:
@gyuho, regarding the logger, today in master, we use `lg != nil` to indicate whether user choose to use zap or plog. Before deprecating plog, we cannot simply provide a no-op `lg` in the top level, right? I think we can keep the `lg != nil` check in this PR. Later when we deprecate plog, we will remove `lg != nil` everywhere.

gyuho commented:
@jingyih Sounds good. Let's just keep it that way for now.

YoyinZyc commented:
Rebased the current master and removed the `lg != nil` check. Tests will be added in the following PR to keep a relative small size of this PR. PHAL.
The failed e2e test seems like a flaky test and it passed locally on my machine. 
```
=== RUN   TestIssue6361
--- FAIL: TestIssue6361 (14.81s)
    ctl_v3_snapshot_test.go:262: read /dev/ptmx: input/output error (expected "val1", got ["{\"level\":\"warn\",\"ts\":\"2020-02-19T23:19:34.677Z\",\"caller\":\"clientv3/retry_interceptor.go:61\",\"msg\":\"retrying of unary invoker failed\",\"target\":\"endpoint://client-6cb9466f-d07c-40ff-bf75-24beea447b31/localhost:20030\",\"attempt\":0,\"error\":\"rpc error: code = DeadlineExceeded desc = context deadline exceeded\"}\r\n" "Error: context deadline exceeded\r\n"])
```
```
=== RUN   TestIssue6361
--- PASS: TestIssue6361 (11.27s)
```

YoyinZyc commented:
@jpbetz Could you please also take a look? Thanks
cc @lavalamp
